src/components/tl/nccl/tl_nccl.h

```diff
@@ -122,17 +126,41 @@ typedef struct ucc_tl_nccl_task {
 UCC_CLASS_DECLARE(ucc_tl_nccl_team_t, ucc_base_context_t *,
                   const ucc_base_team_params_t *);
 
-#define NCCLCHECK_GOTO(_cmd, _label, _status, _lib)                            \
+static inline ucc_status_t ucc_tl_nccl_check_nb(ncclResult_t *nccl_status, // NOLINT
+                                  ucc_status_t *task_st, // NOLINT
+                                  ncclComm_t nccl_comm, //NOLINT
+                                  int check_nb) { //NOLINT
+#if NCCL_USE_NON_BLOCKING
+    if (check_nb &&
+        (*nccl_status == ncclSuccess || *nccl_status == ncclInProgress)) {
+        ncclResult_t st = ncclCommGetAsyncError(nccl_comm, nccl_status);
+        if (st != ncclSuccess) {
+            return UCC_ERR_NO_MESSAGE;
+        }
+        if (ncclInProgress == *nccl_status) {
+            *task_st = UCC_INPROGRESS;
+            return UCC_INPROGRESS;
+        }
+    }
+#endif
+    return UCC_OK;
+}
+
+#define NCCLCHECK_GOTO(_cmd, _label, _st, _lib, _task_st, _comm, _check_nb)    \
```

Comment:
i would rather do something like this

```
do {
    ncclResult_t e, st;
    
    /* check nccl function status */
    e = _cmd;  
    if (ucc_unlikely(e != ncclSuccess && e != ncclInProgress))
    {
        _st = UCC_ERR_NO_MESSAGE;
        goto label;
    }

    /* check nccl comm status */
    st = ncclCommGetAsyncError(nccl_commm, &e);
    if (ucc_unlikely(st != ncclSuccess)) {
        _st = UCC_ERR_NO_MESSAGE;
        goto label;
    }

    if (e == ncclInProgress) {
        _st = UCC_INPROGRESS;
    } else if (e == ncclSuccess) {
        _st = UCC_OK;
    } else {
        _st = UCC_ERR_NO_MESSAGE;
        goto label;
    }
} while(0);
```

Reply:
we can't check if e != ncclInProgress or call ncclCommGetAsyncError unless we are in nccl version that supports nb. This is why I put these checks in seperate function, which if doesn't support nb then will "do nothing" and then back in the macro check if e != ncclSuccess